retrofit/src/main/java/retrofit/RequestFactoryParser.java

```diff
@@ -251,57 +252,121 @@ private void parseParameters(Retrofit retrofit) {
+            } else {
+              Converter<?, String> valueConverter =
+                  retrofit.stringConverter(methodParameterType, methodParameterAnnotations);
+              action = new RequestAction.Query<>(name, valueConverter, encoded);
             }
 
-            action = queryAction;
             gotQuery = true;
 
           } else if (methodParameterAnnotation instanceof QueryMap) {
             if (!Map.class.isAssignableFrom(Utils.getRawType(methodParameterType))) {
               throw parameterError(i, "@QueryMap parameter type must be Map.");
             }
+            if (!(methodParameterType instanceof ParameterizedType)) {
+              throw parameterError(i, "Map must include generic types (e.g., Map<String, String>)");
+            }
+            ParameterizedType parameterizedType = (ParameterizedType) methodParameterType;
+            Type keyType = Utils.getParameterUpperBound(0, parameterizedType);
+            if (String.class != keyType) {
+              throw parameterError(i, "@QueryMap keys must be of type String: " + keyType);
+            }
+            Type valueType = Utils.getParameterUpperBound(1, parameterizedType);
+            Converter<?, String> valueConverter =
+                retrofit.stringConverter(valueType, methodParameterAnnotations);
+
             QueryMap queryMap = (QueryMap) methodParameterAnnotation;
-            action = new RequestAction.QueryMap(queryMap.encoded());
+            action = new RequestAction.QueryMap<>(valueConverter, queryMap.encoded());
 
           } else if (methodParameterAnnotation instanceof Header) {
             Header header = (Header) methodParameterAnnotation;
+            String name = header.value();
 
             Class<?> rawParameterType = Utils.getRawType(methodParameterType);
-            RequestAction<?> headerAction = new RequestAction.Header(header.value());
             if (Iterable.class.isAssignableFrom(rawParameterType)) {
-              headerAction = headerAction.iterable();
+              if (!(methodParameterType instanceof ParameterizedType)) {
+                throw parameterError(i, rawParameterType.getSimpleName()
+                    + " must include generic type (e.g., "
+                    + rawParameterType.getSimpleName()
+                    + "<String>)");
+              }
+              ParameterizedType parameterizedType = (ParameterizedType) methodParameterType;
+              Type iterableType = Utils.getParameterUpperBound(0, parameterizedType);
+              Converter<?, String> valueConverter =
+                  retrofit.stringConverter(iterableType, methodParameterAnnotations);
+              action = new RequestAction.Header<>(name, valueConverter).iterable();
             } else if (rawParameterType.isArray()) {
-              headerAction = headerAction.array();
+              Class<?> arrayType = rawParameterType.getComponentType(); // TODO box
```

Comment:
Punting that choice & testing it: https://github.com/square/retrofit/issues/1301